File: backend/repositories/shiftRepository.js
import db from '../config/db.js';

/**
 * Get all shifts
 * @returns {Promise<Array>}
 */
export const getAllShifts = async () => {
  const [rows] = await db.query('SELECT * FROM shifts ORDER BY name ASC');
  return rows;
};

/**
 * Get shift by ID
 * @param {number} id
 * @returns {Promise<Object|null>}
 */
export const getShiftById = async (id) => {
  const [rows] = await db.query('SELECT * FROM shifts WHERE id = ?', [id]);
  return rows[0] || null;
};

/**
 * Create new shift
 * @param {Object} data
 * @returns {Promise<number>} Inserted ID
 */
export const createShift = async (data) => {
  const { name, start_time, end_time, work_days, flexible_minutes, is_default } = data;
  const [result] = await db.query(
    'INSERT INTO shifts (name, start_time, end_time, work_days, flexible_minutes, is_default) VALUES (?, ?, ?, ?, ?, ?)',
    [name, start_time, end_time, work_days, flexible_minutes, is_default]
  );
  return result.insertId;
};

/**
 * Update shift
 * @param {number} id
 * @param {Object} data
 * @returns {Promise<number>} Affected Rows
 */
export const updateShift = async (id, data) => {
  const { name, start_time, end_time, work_days, flexible_minutes, is_default } = data;
  const [result] = await db.query(
    'UPDATE shifts SET name = ?, start_time = ?, end_time = ?, work_days = ?, flexible_minutes = ?, is_default = ? WHERE id = ?',
    [name, start_time, end_time, work_days, flexible_minutes, is_default, id]
  );
  return result.affectedRows;
};

/**
 * Delete shift
 * @param {number} id
 * @returns {Promise<number>} Affected Rows
 */
export const deleteShift = async (id) => {
  const [result] = await db.query('DELETE FROM shifts WHERE id = ?', [id]);
  return result.affectedRows;
};

/**
 * Get default shift (fallback)
 * @returns {Promise<Object>}
 */
export const getDefaultShift = async () => {
  const [rows] = await db.query('SELECT * FROM shifts WHERE is_default = 1 LIMIT 1');
  return rows[0] || null;
};

/**
 * Get shift for specific user, fallback to default if not assigned.
 * @param {string} usernameOrId
 * @returns {Promise<Object>}
 */
export const getUserShift = async (username) => {
  // Priority: User's assigned shift > Default shift
  const query = `
        SELECT s.*
        FROM users u
        LEFT JOIN shifts s ON u.shift_id = s.id
        WHERE u.username = ?
    `;
  const [rows] = await db.query(query, [username]);

  if (rows.length > 0 && rows[0].id) {
    return rows[0];
  }
  return await getDefaultShift();
};

/**
 * Get User ID by Username
 * @param {string} username
 * @returns {Promise<number|null>}
 */
export const getUserIdByUsername = async (username) => {
  const [rows] = await db.query('SELECT id FROM users WHERE username = ?', [username]);
  return rows[0]?.id || null;
};

File: backend/internal/repository/shift_repository.go
package repository

import (
	"context"

	"github.com/dps-wmhris/backend/internal/model"
	"github.com/jmoiron/sqlx"
)

// ShiftRepository defines the interface for shift data operations
type ShiftRepository interface {
	GetAll(ctx context.Context) ([]model.Shift, error)
	GetByID(ctx context.Context, id int) (*model.Shift, error)
	Create(ctx context.Context, shift *model.Shift) (int, error)
	Update(ctx context.Context, shift *model.Shift) error
	Delete(ctx context.Context, id int) error
	ClearDefault(ctx context.Context, ext sqlx.ExtContext) error
	GetUserShift(ctx context.Context, username string) (*model.Shift, error)
}

type shiftRepositoryImpl struct {
	db *sqlx.DB
}

// NewShiftRepository injects the database dependency
func NewShiftRepository(db *sqlx.DB) ShiftRepository {
	return &shiftRepositoryImpl{db: db}
}

func (r *shiftRepositoryImpl) GetAll(ctx context.Context) ([]model.Shift, error) {
	var shifts []model.Shift
	query := `SELECT * FROM shifts ORDER BY name ASC`
	err := r.db.SelectContext(ctx, &shifts, query)
	if err != nil {
		return nil, err
	}
	return shifts, nil
}

func (r *shiftRepositoryImpl) GetByID(ctx context.Context, id int) (*model.Shift, error) {
	var shift model.Shift
	query := `SELECT * FROM shifts WHERE id = ?`
	err := r.db.GetContext(ctx, &shift, query, id)
	if err != nil {
		return nil, err
	}
	return &shift, nil
}

func (r *shiftRepositoryImpl) Create(ctx context.Context, shift *model.Shift) (int, error) {
	query := `
		INSERT INTO shifts (name, start_time, end_time, work_days, flexible_minutes, is_default) 
		VALUES (?, ?, ?, ?, ?, ?)`
	result, err := r.db.ExecContext(ctx, query,
		shift.Name, shift.StartTime, shift.EndTime, shift.WorkDays, shift.FlexibleMinutes, shift.IsDefault,
	)
	if err != nil {
		return 0, err
	}
	id, err := result.LastInsertId()
	if err != nil {
		return 0, err
	}
	return int(id), nil
}

func (r *shiftRepositoryImpl) Update(ctx context.Context, shift *model.Shift) error {
	query := `
		UPDATE shifts 
		SET name = ?, start_time = ?, end_time = ?, work_days = ?, flexible_minutes = ?, is_default = ? 
		WHERE id = ?`
	_, err := r.db.ExecContext(ctx, query,
		shift.Name, shift.StartTime, shift.EndTime, shift.WorkDays, shift.FlexibleMinutes, shift.IsDefault, shift.ID,
	)
	return err
}

func (r *shiftRepositoryImpl) Delete(ctx context.Context, id int) error {
	query := `DELETE FROM shifts WHERE id = ?`
	_, err := r.db.ExecContext(ctx, query, id)
	return err
}

// ClearDefault resets all shifts' is_default to false. Used transactionally.
func (r *shiftRepositoryImpl) ClearDefault(ctx context.Context, ext sqlx.ExtContext) error {
	if ext == nil {
		ext = r.db
	}
	query := `UPDATE shifts SET is_default = false`
	_, err := ext.ExecContext(ctx, query)
	return err
}

func (r *shiftRepositoryImpl) GetUserShift(ctx context.Context, username string) (*model.Shift, error) {
	var shift model.Shift
	// Priority 1: User's assigned shift
	queryUserShift := `
		SELECT s.* 
		FROM users u 
		LEFT JOIN shifts s ON u.shift_id = s.id 
		WHERE u.username = ?`
	
	err := r.db.GetContext(ctx, &shift, queryUserShift, username)
	if err == nil && shift.ID != 0 {
		return &shift, nil
	}

	// Priority 2: Default shift
	queryDefault := `SELECT * FROM shifts WHERE is_default = 1 LIMIT 1`
	err = r.db.GetContext(ctx, &shift, queryDefault)
	if err != nil {
		return nil, err
	}

	return &shift, nil
}
